### PR TITLE
Draft: Use crane instead of naersk

### LINKED ISSUE #483: Replace naersk by crane
Naersk owner seems to redirect users to [crane](https://github.com/ipetkov/crane), see https://github.com/nix-community/naersk/pull/214#issuecomment-1963774309.

### PR DESCRIPTION
**Pull request by KoviRobi**
Fixes #483

I thought I'd have a try at doing this, also made the package more normal package like, and doing the artifact building in the CI (though to keep the zip/tar paths same the command is not super great).

I also want to have a look at adding feature flags, though can't get `withNotmuch` to build yet, it can't find `-lnotmuch`, it doesn't seem to end up in the libs for some reason, though it does start out in `NIX_LDFLAGS`.

### CODE REVIEW

**Review by soywod — CHANGES_REQUESTED** [review]
Thank you so much, I wanted to tackle this task soon. I have many things to say:

- Actually there are many tasks related to Nix: crane switch, overriding, and share common code with nixpkgs. Let's focus on the crane switch with this PR.
- What is your opinion about fenix vs mozilla overlay? Both does not seem to be so active.
- I find overriding not intuitive at all with flakes. My ultimate aim is to be able to configure apps, packages and devShells based on cargo features. For example, the notmuch feature is disabled by default. The notmuch lib should not be present in the buildInputs of the devShell nor the cargo features. But if someone needs the feature, this person should be able to easily have a dev shell with libnotmuch and notmuch cargo feature enabled. Is there a convenient way to do so?
- A [nixpkgs derivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/hi/himalaya/package.nix) already exists for Himalaya CLI, and it clearly duplicates with the actual flake. What do you advise?
  - Make the nixpkgs derivation the source of truth, then import it from the flake.
  - Make the flake the source of truth, then import it from nixpkgs.
  - Implement a shared interface, so both use the same thing.

File: .github/workflows/pre-release.yml [hunk 1]
```diff
@@ -39,10 +39,26 @@ jobs:
...
+          tar -cf $OLDPWD/himalaya.tar himalaya*
+          cd -
+          cd result
+          tar -rf $OLDPWD/himalaya.tar share
+          cd -
+          gzip himalaya.tar
+          mv himalaya.tar.gz himalaya.tgz
+
+          cd result/bin
+          zip -r $OLDPWD/himalaya.zip himalaya*
```
[the remark this answers is not in the document]
Comment (KoviRobi) [reply .github/workflows/pre-release.yml:53]: Fair, I'm happy to move this to be considered later, especially after we have a better idea of what should be the source of truth you asked above

File: flake.nix [hunk 2]
```diff
@@ -13,17 +13,10 @@
       url = "github:soywod/fenix";
       inputs.nixpkgs.follows = "nixpkgs";
     };
-    naersk = {
-      url = "github:nix-community/naersk";
-      inputs.nixpkgs.follows = "nixpkgs";
-    };
-    flake-compat = {
```
[the remark this answers is not in the document]
Comment (KoviRobi) [reply flake.nix:20]: Ah I just thought it wasn't used, my bad

File: flake.nix [hunk 3]
```diff
@@ -72,11 +65,6 @@
 
         x86_64-darwin.x86_64-darwin = {
           rustTarget = "x86_64-apple-darwin";
-          mkPackage = { pkgs, ... }: package:
-            let inherit (pkgs.darwin.apple_sdk_11_0.frameworks) Security;
-            in package // {
-              NIX_LDFLAGS = "-F${Security}/Library/Frameworks -framework Security";
```
[the remark this answers is not in the document]
Comment (KoviRobi) [reply flake.nix:78]: I think just adding it to the build inputs does the same. The thing about using the attrset `//` is that it doesn't merge nested attributes, so it makes overriding more difficult.

**Comment by KoviRobi** [comment]
Hmm for some reason github isn't letting me reply on this one

> * Actually there are many tasks related to Nix: crane switch, overriding, and share common code with nixpkgs. Let's focus on the crane switch with this PR.
> * What is your opinion about fenix vs mozilla overlay? Both does not seem to be so active.

I'm not that familiar with the Rust ecosystem to know the ins and outs, I just use https://github.com/oxalica/rust-overlay for my projects because that seemed simplest

> * I find overriding not intuitive at all with flakes. My ultimate aim is to be able to configure apps, packages and devShells based on cargo features. For example, the notmuch feature is disabled by default. The notmuch lib should not be present in the buildInputs of the devShell nor the cargo features. But if someone needs the feature, this person should be able to easily have a dev shell with libnotmuch and notmuch cargo feature enabled. Is there a convenient way to do so?

Definitely agree it's awkward, the whole `---expr getFlake overrideAttrs` I had to do isn't very readable or nice. It would be much simpler if we could do `nix build/develop . --override withNotmuch true` or even just `nix build .#default.override --args withNotmuch true`. Alas AFAIK currently this is not supported in flakes, which I think is an oversight.

I understand why flakes themselves want to be pure for caching, but I think it would be great if we could have commands/env vars as inputs to the flake.

> * A [nixpkgs derivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/hi/himalaya/package.nix) already exists for Himalaya CLI, and it clearly duplicates with the actual flake. What do you advise?
>   
>   1. Make the nixpkgs derivation the source of truth, then import it from the flake.
>   2. Make the flake the source of truth, then import it from nixpkgs.
>   3. Implement a shared interface, so both use the same thing.

I think 1 will end up having a lot of maintenance burden if you want to change something. I think 2 is probably the best, as you can continue to develop the flake, modify it as you need it, and if you need to update it in nixpkgs then it's just a simple bump changing the ref/hash. The difference between 2 and 3 is I think just a matter of if you want to have somethings special to the flake or not.

**Comment by soywod** [comment]
> I think 2 is probably the best, as you can continue to develop the flake, modify it as you need it, and if you need to update it in nixpkgs then it's just a simple bump changing the ref/hash.

I agree, especially that nixpkgs is less accessible. Better to do the minimum there.

> The difference between 2 and 3 is I think just a matter of if you want to have somethings special to the flake or not.

The idea behind is to have shared stuff in nix files. It could help us, for example, to use `--override-flake` or `--override-input` more easily, it could be re-used for other Pimalaya projects etc.

---

I approve the PR, it can be merged! Thank you for the changes.

**Comment by KoviRobi** [comment]
> The idea behind is to have shared stuff in nix files. It could help us, for example, to use `--override-flake` or `--override-input` more easily, it could be re-used for other Pimalaya projects etc.

Ah I see, yeah it would be nice to share what is essentially almost the same flake between himalaya/mirador/neverest

**Comment by soywod** [comment]
It would also interface easier with nixpkgs, as they seem to use a different Rust platform. Feel free to propose the override in another PR, meanwhile I will experiment a bit how to nicely customize cargo features via flakes.

**Pull request merged**